File: str/reverse/reverseWordInSentence1.java
//no heading or ending white spaces
//no duplicate white space
//eg: “I love Google” → “Google love I”
public String reverseWords(String s) {
        // write your code here
  if(s == null || s.length() == 0)
		return "";
	String[] arr = s.split(" ");
	StringBuilder sb = new StringBuilder();
	for(int i = arr.length -1; i >=0; i--){
		if(!arr[i].equals(" "))
			sb.append(arr[i]).append(" ");
	}
	//if input is " "
	if(sb.length() == 0)
		return "";
	else //remove the last " " from line 11
		return sb.substring(0, sb.length()-1);
		//cant use sb.toString()
    }

//solution2: O(n), O(1)
//can handle heading and ending whitespace, but not duplciate space
//similar to remove duplicate white space
public Stirng reverseWords(String input) {
if(input == null || input.length <= 1)
      return input;
    char[] arr = input.toCharArray();
    int start = 0;
    //int fast = 0;
    //reverse each word=> i evol     oohay
    //reverse(arr, 0, arr.length -1);
   	for(int fast = 0; fast < arr.length; fast++){
      //start index of a word
      if(arr[fast] != ' ' && (fast == 0 || arr[fast-1]== ' '))
         start = fast;
      //end index of a word
      if(arr[fast] != ' ' && (fast == arr.length -1 || arr[fast+1] == ' '))
        reverse(arr, start, fast);
    }
    //reverse the senctence => i yahoo     love i
    reverse(arr, 0, arr.length -1);
    return new String(arr);
  }
  // i love     yahoo
  // i love     yahoo
  // i love     yahoo
  // i love     yahoo
  // i love     yahoo
  // i evol     yahoo
  // i  love    yahoo
  // i  evol    yahoo
  // i   love   yahoo
  // i   evol   yahoo
  // i   evol   yahoo
  // i   evol   yahoo
  // i   evol   yahoo
  // i   evol   yahoo
  // i   evol   yahoo
  // i   evol   oohay
  // yahoo   love   i

  private static void reverse(char[] arr, int left, int right){
    if(left > right){
    	return;
    }
    swap(arr, left, right);
    reverse(arr, left+1, right-1);
  }
  private void swap(char[] arr, int left, int right){
  	char temp = arr[left];
    arr[left] = arr[right];
    arr[right] = temp;
  }

File: str/reverse/reverseWordInSentence1.py
# 1. Reverse the entire string.
# 2. Reverse each word.
# 3. Deduplicate spaces.

# can not handle leading and tail white space
def reverseWords(self, s: str) -> str:
#         res = ' '.join(x[::-1] for x in s.split())
#         return ''.join(res[::-1])
        # return ' '.join(reversed(s.split()))
    def removeSpace(arr):
        cur = 0
        for i, item in enumerate(arr):
            if arr[i] == ' ' and (i == 0 or arr[i-1] == ' '):
                continue
            arr[cur] = arr[i]
            cur += 1
        return arr[0:cur-1] if cur > 0 and arr[cur-1] == ' ' else arr[0:cur]

    def reverse_range(arr, left, right):
        while left < right:
            arr[left],arr[right] = arr[right], arr[left]
            left += 1
            right -= 1

    arr = removeSpace(list(s))

    start = 0
    for i, item in enumerate(arr):
        if arr[i] != ' ' and (i == 0 or arr[i-1] == ' '):
            start = i
        if arr[i] != ' ' and (i==len(arr)-1 or arr[i+1] == ' '):
            reverse_range(arr, start, i)

    reverse_range(arr, 0, len(arr)-1)

    return ''.join(arr)

# follow up1:
'''
Input:  ["t","h","e"," ","s","k","y"," ","i","s"," ","b","l","u","e"]
Output: ["b","l","u","e"," ","i","s"," ","s","k","y"," ","t","h","e"]
'''
    s[:] = ' '.join(reversed(''.join(s).split()))
